### ISSUE TITLE
.desktop shortcuts are printed in different language

### ISSUE THREAD

**Issue by ghost** [davatorium/rofi#938]
## Version
Version: 1.5.2

## Configuration
[rofi -help](https://gist.github.com/chzzzh/c6271ca3edf119789b53bb432fc9f13b)

## Launch Command
rofi -show drun

## Steps to reproduce
Just run Rofi.

## What behaviour you see
Rofi shows menu items in Russian language.

## What behaviour you expect to see
To see them in English.

## $ locale
```
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC=ru_UA.UTF-8
LC_TIME=ru_UA.UTF-8
LC_COLLATE=en_US.UTF-8
LC_MONETARY=ru_UA.UTF-8
LC_MESSAGES="en_US.UTF-8"
LC_PAPER=ru_UA.UTF-8
LC_NAME=ru_UA.UTF-8
LC_ADDRESS=ru_UA.UTF-8
LC_TELEPHONE=ru_UA.UTF-8
LC_MEASUREMENT=ru_UA.UTF-8
LC_IDENTIFICATION=ru_UA.UTF-8
LC_ALL=
```

P. S.: changing every LC_* to 'en_US.UTF-8' didn't help previously.

xoxo to every contributor 🖤

**Comment by DaveDavenport** [comment]
Odd, I do not handle that myself. I let GLIB sort it out.
on startup I load the locales

```
setlocale(LC_ALL, "" );
``` (see man setlocale)

and for the reset `g_key_file_get_locale_string`.

If you set *all* to en_US, it should not show Russian.

**Comment by chzzzh** [comment]
Isn't it better to use the fallback locale after LC_ALL? Not really the nicest way out to put all the locales to the only one.

**Comment by sardemff7** [comment]
I can confirm by testing with `env LC_MESSAGES=en_US.UTF-8 rofi -show drun` and `env LC_MESSAGES=fr_FR.UTF-8 rofi -show drun` that it works as expected.
`setlocale(LC_ALL, "");` is the way to set all `LC_*` according to this order of env: `LC_ALL`, `LC_<category>` then `LANG`. First to be set is used.

99% of env issues are entirely non-rofi issues.

You have to make sure rofi has the correct env. Did you check that?

**Comment by chzzzh** [comment]
So I generated extra `fr_FR.UTF-8`, but both `env LC_MESSAGES=fr_FR.UTF-8 rofi -show drun` and `env LC_MESSAGES=en_US.UTF-8 rofi -show drun` are showing Rofi in drun-mode with Russian options. I checked `~/.local/share/applications`, but there are almost no shortcuts, everything's ok with `/usr/share/applications`. I'm in confusion.

**Comment by sardemff7** [comment]
Is it possible that your system only installed the Russian translations in your `.desktop` files?
I tried something a little bit different and `env LC_MESSAGES=en_US.UTF-8 rofi -show drun` is actually using the fallback locale (`C`) on my system because very few of my `.desktop` files actually have and `en_US`-specific value.
You could try `env LC_MESSAGES=C rofi -show drun` but I’d guess it won’t work because the issue is elsewhere.

At this point it would help to know more about your system, which distribution, which variant, which version it is. Also, could you share with us the content of one of your `.desktop` files?

**Comment by chzzzh** [comment]
@sardemff7, thanks for the tip! As I stated before, there's everything ok with `.desktop` shortcuts - most, if not all, are filled with defaults and many other languages' variables. `env LC_MESSAGES=C rofi -show drun` didn't give a result as well.

So I just found out, after changing `LANGUAGE` locale from [`en:C:ru`](https://unix.stackexchange.com/a/181867/320683) only to `en` and rebooting everything worked out! Respectively, `env LANGUAGE=ru rofi -show drun` showed Russian variables.

**Comment by DaveDavenport** [comment]
rofi does not implement any locale code itself, it relies on 'setlocale' and glib for this.
If the behaviour is not correct, please report a bug upstream.

**Comment by chzzzh** [comment]
@DaveDavenport, I'm reporting, that Rofi relies on last language from all the values in the `LANGUAGE` variable, instead of the first one 🙂

**Comment by DaveDavenport** [comment]
And I am saying that rofi does not rely on the `LANGUAGE` variable and that most likely an upstream library does.

**Comment by sardemff7** [comment]
Again, rofi doesn’t read this variable. GLib does, in `g_get_language_names()` which is used by its `GKeyFile` API. rofi **then** use the `GKeyFile` API, as expected, to read `.desktop` file. If the `LANGUAGE` variable is used wrong, it is either in `g_get_language_names()` when reading it or in `GKeyFile` when using the read values.

I searched the code a bit, I can see one possible reason: the locale `C` is not special-cased in `g_key_file_get_locale_string()`, but only used as a callback. I would test that: copy one `.desktop` file to `~/.local/share/applications`, then edit it and for every key with locale-specific value (e.g. `Comment[ru]=`) add the default one with `[C]`. If it works, then we found the issue.
However, I’m not sure it is an error in any way. The `C` locale **is** meant to be a fallback value, though sadly often used for `en_US`, lazily, probably. :-)
But I would report it to GLib, or at least check if they have any report mentionning it, as it might be a `WONTFIX` to them. :-)
